wen-restart/src/wen_restart.rs

```diff
@@ -126,26 +343,97 @@ mod tests {
+        let bank_forks_clone = bank_forks.clone();
+        let wen_restart_thread_handle = Builder::new()
+            .name("solana-wen-restart".to_string())
+            .spawn(move || {
+                assert!(wait_for_wen_restart(
+                    &wen_restart_proto_path_clone,
+                    VoteTransaction::from(Vote::new(vec![last_vote_slot], last_vote_bankhash)),
+                    blockstore,
+                    cluster_info_clone,
+                    bank_forks_clone,
+                    wen_restart_repair_slots.clone(),
+                    80,
+                    Arc::new(AtomicBool::new(false)),
+                )
+                .is_ok());
+            })
+            .unwrap();
+        let mut rng = rand::thread_rng();
+        let mut expected_messages = HashMap::new();
+        // Skip the first 2 validators, because 0 is myself, we only need 8 more to reach > 80%.
+        for keypairs in validator_voting_keypairs.iter().skip(2) {
+            let node_pubkey = keypairs.node_keypair.pubkey();
+            let node = LegacyContactInfo::new_rand(&mut rng, Some(node_pubkey));
+            let last_vote_hash = Hash::new_unique();
+            let slots = RestartLastVotedForkSlots::new(
+                node_pubkey,
+                timestamp(),
+                &expected_slots_to_repair,
+                last_vote_hash,
+                shred_version,
+            )
+            .unwrap();
+            let entries = vec![
+                CrdsValue::new_signed(CrdsData::LegacyContactInfo(node), &keypairs.node_keypair),
+                CrdsValue::new_signed(
+                    CrdsData::RestartLastVotedForkSlots(slots),
+                    &keypairs.node_keypair,
+                ),
+            ];
+            {
+                let mut gossip_crds = cluster_info.gossip.crds.write().unwrap();
+                for entry in entries {
+                    assert!(gossip_crds
+                        .insert(entry, /*now=*/ 0, GossipRoute::LocalMessage)
+                        .is_ok());
+                }
+            }
+            expected_messages.insert(
+                node_pubkey.to_string(),
+                LastVotedForkSlotsRecord {
+                    last_vote_fork_slots: expected_slots_to_repair.clone(),
+                    last_vote_bankhash: last_vote_hash.to_string(),
+                    shred_version: shred_version as u32,
+                },
+            );
+        }
+        let mut parent_bank = bank_forks.read().unwrap().root_bank();
+        for slot in expected_slots_to_repair {
+            let mut my_bank_forks = bank_forks.write().unwrap();
+            my_bank_forks.insert(Bank::new_from_parent(
+                parent_bank.clone(),
+                &Pubkey::default(),
+                slot,
+            ));
+            parent_bank = my_bank_forks.get(slot).unwrap();
+            parent_bank.freeze();
+        }
+        let _ = wen_restart_thread_handle.join();
```

Comment:
Because we're essentially testing the `RestartState` state machine, I would like to see a more comprehensive test that tests all the combinations of states/failures.

I think a structure like the following might make sense where we have a `begin()`, `failure()` and `finish()` method for simulating a scenario for each of the states? Feel free to adapt this as you see fit:

```
fn test_wen_restart(failing_states: HashMap<RestartState, bool> // list of all states we want to fail) {
  let validator_state = init();

  // Begin sending last voted fork slots
  begin_last_voted_fork_slots(&mut validator_state);
  if failing_states.get(RestartState::LastVotedForkSlots) {
       incur_last_voted_forks_failure(&mut validator_state);
       restart_validator();
  }
  finish_last_voted_fork_slots(&mut validator_state);

   // Repeat for the other RestartState's


   // Check we get to Done
}

```

Reply:
That's a good point.

Since wait_for_wen_restart uses a loop, it's very hard to test it one step at a time. I added some failure tests and a separate test for state transition, does this look better?